sync/sync.go

```diff
@@ -1,78 +1,36 @@
 package sync
 
 import (
-	"github.com/go-kit/kit/log"
-	"github.com/pkg/errors"
-
 	"github.com/weaveworks/flux/cluster"
-	"github.com/weaveworks/flux/policy"
 	"github.com/weaveworks/flux/resource"
 )
 
-// Sync synchronises the cluster to the files in a directory
-func Sync(logger log.Logger, m cluster.Manifests, repoResources map[string]resource.Resource, clus cluster.Cluster,
-	deletes bool) error {
-	// Get a map of resources defined in the cluster
-	clusterBytes, err := clus.Export()
-
-	if err != nil {
-		return errors.Wrap(err, "exporting resource defs from cluster")
-	}
-	clusterResources, err := m.ParseManifests(clusterBytes)
-	if err != nil {
-		return errors.Wrap(err, "parsing exported resources")
-	}
-
-	// Everything that's in the cluster but not in the repo, delete;
-	// everything that's in the repo, apply. This is an approximation
-	// to figuring out what's changed, and applying that. We're
-	// relying on Kubernetes to decide for each application if it is a
-	// no-op.
-	sync := cluster.SyncDef{}
+// Syncer has the methods we need to be able to compile and run a sync
+type Syncer interface {
+	Sync(cluster.SyncDef) error
+}
 
-	// DANGER ZONE (tamara) This works and is dangerous. At the moment will delete Flux and
-	// other pods unless the relevant manifests are part of the user repo. Needs a lot of thought
-	// before this cleanup cluster feature can be unleashed on the world.
-	if deletes {
-		for id, res := range clusterResources {
-			prepareSyncDelete(logger, repoResources, id, res, &sync)
-		}
-	}
+// Sync synchronises the cluster to the files under a directory.
+func Sync(repoResources map[string]resource.Resource, clus Syncer) error {
+	// TODO: multiple stack support. This will involve partitioning
+	// the resources into disjoint maps, then passing each to
+	// makeStack.
+	defaultStack := makeStack("default", repoResources)
```

Comment:
Ah! So stacks are not really used yet. Isn't it a bit of over-engineering to add all the infrastructure before they are actually used?

Reply:
A little, yes -- it amounts to a couple of unnecessary loops, and a bit of cognitive burden. If I were to remove them, I'd just strip away the outer struct and the loops.

Reply:
I would vote for removing them for now. My reasoning is:
* This PR doesn't make use of multiple stacks yet.
* Sounds like a premature optimization. Do we really need them? Since we need to load the resources into memory for parsing anyways, I doubt that checksumming of each resource would be too expensive.

Reply:
> * I doubt that checksumming of each resource would be too expensive.

It's done per resource anyway